File: server/Api/Store.cs
using Microsoft.EntityFrameworkCore;

namespace Api;

/// <summary>
/// The queries, and nothing else: no connection, no HTTP. Each takes the
/// context as its first argument so the same code runs against the API's
/// Postgres and against the one the tests start (see ../Api.Tests/StoreTests.cs).
/// </summary>
public static class Store
{
    public static async Task<Account?> FindAccount(AccountsDb db, string session)
    {
        var row = await db.Accounts.AsNoTracking().SingleOrDefaultAsync(row => row.Session == session);
        return row?.ToAccount();
    }

    /// <summary>Stores the account for this session, replacing whatever was there.</summary>
    public static async Task<Account> SaveAccount(AccountsDb db, string session, Account account)
    {
        var row = await db.Accounts.FindAsync(session);
        if (row is null)
        {
            row = new AccountRow
            {
                Session = session,
                Id = account.Id,
                Nom = account.Nom,
                Prenom = account.Prenom,
                Email = account.Email,
                Telephone = account.Telephone,
                Langue = account.Langue,
                Bio = account.Bio,
            };
            db.Accounts.Add(row);
        }
        else
        {
            row.Id = account.Id;
            row.Apply(account);
        }
        await db.SaveChangesAsync();
        return row.ToAccount();
    }

    /// <summary>
    /// Replaces every field the client may send, and keeps the id: the server
    /// owns it, so it is not in the payload and cannot be overwritten by one.
    /// Null when the session has no account to update.
    /// </summary>
    public static async Task<Account?> UpdateAccount(AccountsDb db, string session, AccountPayload payload)
    {
        var row = await db.Accounts.FindAsync(session);
        if (row is null)
        {
            return null;
        }
        row.Apply(payload);
        await db.SaveChangesAsync();
        return row.ToAccount();
    }
}

File: server/store.ts
/**
 * The queries, and nothing else: no connection, no HTTP. Each takes the
 * database as its first argument so the same code runs against Postgres in the
 * API and against PGlite in the unit tests (see __tests__/store.unit.test.ts).
 */
import { eq } from "drizzle-orm";
import type { PgDatabase, PgQueryResultHKT } from "drizzle-orm/pg-core";
import type { Account, AccountPayload } from "../src/account/helpers/api.ts";
import { accounts } from "./schema.ts";

/** Any Drizzle Postgres database, whichever driver is behind it. */
export type Database = PgDatabase<PgQueryResultHKT>;

/** Every column but the session: what goes back over the wire. */
const accountColumns = {
  id: accounts.id,
  nom: accounts.nom,
  prenom: accounts.prenom,
  email: accounts.email,
  telephone: accounts.telephone,
  langue: accounts.langue,
  bio: accounts.bio,
};

export async function findAccount(db: Database, session: string): Promise<Account | null> {
  const [found] = await db
    .select(accountColumns)
    .from(accounts)
    .where(eq(accounts.session, session));
  return found ?? null;
}

/** Stores the account for this session, replacing whatever was there. */
export async function saveAccount(
  db: Database,
  session: string,
  account: Account,
): Promise<Account> {
  const [saved] = await db
    .insert(accounts)
    .values({ session, ...account })
    .onConflictDoUpdate({ target: accounts.session, set: account })
    .returning(accountColumns);
  return saved;
}

/**
 * Replaces every field the client may send, and keeps the id: the server owns
 * it, so it is not in the payload and cannot be overwritten by one. Resolves
 * null when the session has no account to update.
 */
export async function updateAccount(
  db: Database,
  session: string,
  payload: AccountPayload,
): Promise<Account | null> {
  const [updated] = await db
    .update(accounts)
    .set(payload)
    .where(eq(accounts.session, session))
    .returning(accountColumns);
  return updated ?? null;
}
